File: Aula 07 - Revisão da Prova/Go/AbstractFactory.go
package main

import "fmt"

// Interface Abstract Factory
type CarFactory interface {
    CreateSedan() Sedan
    CreateSUV() SUV
}

// Implementação da fábrica para Carros Toyota
type ToyotaFactory struct{}

func (t *ToyotaFactory) CreateSedan() Sedan {
    return &ToyotaSedan{}
}

func (t *ToyotaFactory) CreateSUV() SUV {
    return &ToyotaSUV{}
}

// Implementação da fábrica para Carros Ford
type FordFactory struct{}

func (f *FordFactory) CreateSedan() Sedan {
    return &FordSedan{}
}

func (f *FordFactory) CreateSUV() SUV {
    return &FordSUV{}
}

// Interface para os produtos
type Sedan interface {
    DisplayInfo()
}

type SUV interface {
    DisplayInfo()
}

// Implementações de produtos específicos para Toyota
type ToyotaSedan struct{}

func (s *ToyotaSedan) DisplayInfo() {
    fmt.Println("Toyota Corolla - Sedan")
}

type ToyotaSUV struct{}

func (s *ToyotaSUV) DisplayInfo() {
    fmt.Println("Toyota RAV4 - SUV")
}

// Implementações de produtos específicos para Ford
type FordSedan struct{}

func (s *FordSedan) DisplayInfo() {
    fmt.Println("Ford Fusion - Sedan")
}

type FordSUV struct{}

func (s *FordSUV) DisplayInfo() {
    fmt.Println("Ford Explorer - SUV")
}

// Função que utiliza a fábrica para criar carros
func produceCars(factory CarFactory) {
    sedan := factory.CreateSedan()
    suv := factory.CreateSUV()

    sedan.DisplayInfo()
    suv.DisplayInfo()
}

func main() {
    fmt.Println("Produzindo carros Toyota:")
    toyotaFactory := &ToyotaFactory{}
    produceCars(toyotaFactory)

    fmt.Println("\nProduzindo carros Ford:")
    fordFactory := &FordFactory{}
    produceCars(fordFactory)
}

File: Aula 07 - Revisão da Prova/Java/AbstractFactory.java
// Interface Abstract Factory
interface CarFactory {
    Sedan createSedan();
    SUV createSUV();
}

// Implementação da fábrica para Carros Toyota
class ToyotaFactory implements CarFactory {
    public Sedan createSedan() {
        return new ToyotaSedan();
    }

    public SUV createSUV() {
        return new ToyotaSUV();
    }
}

// Implementação da fábrica para Carros Ford
class FordFactory implements CarFactory {
    public Sedan createSedan() {
        return new FordSedan();
    }

    public SUV createSUV() {
        return new FordSUV();
    }
}

// Interface para os produtos
interface Sedan {
    void displayInfo();
}

interface SUV {
    void displayInfo();
}

// Implementações de produtos específicos para Toyota
class ToyotaSedan implements Sedan {
    public void displayInfo() {
        System.out.println("Toyota Corolla - Sedan");
    }
}

class ToyotaSUV implements SUV {
    public void displayInfo() {
        System.out.println("Toyota RAV4 - SUV");
    }
}

// Implementações de produtos específicos para Ford
class FordSedan implements Sedan {
    public void displayInfo() {
        System.out.println("Ford Fusion - Sedan");
    }
}

class FordSUV implements SUV {
    public void displayInfo() {
        System.out.println("Ford Explorer - SUV");
    }
}

// Função que utiliza a fábrica para criar carros
public class Main {
    public static void produceCars(CarFactory factory) {
        Sedan sedan = factory.createSedan();
        SUV suv = factory.createSUV();

        sedan.displayInfo();
        suv.displayInfo();
    }

    public static void AbstractFactory(String[] args) {
        System.out.println("Produzindo carros Toyota:");
        CarFactory toyotaFactory = new ToyotaFactory();
        produceCars(toyotaFactory);

        System.out.println("\nProduzindo carros Ford:");
        CarFactory fordFactory = new FordFactory();
        produceCars(fordFactory);
    }
}
